In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('/Users/raidneghiz/Documents/ydata-profiling2/Tunisair_flights_dataset.csv' ,
                 sep = ',' , header= 0 , 
                 names = ['Filght_date','Flight_ID','Departure_point','Arrival_point','Scheduled_departure_time','Scheduled_arrival_time','STATUS','Aircraft_code','Arrival_delay'],
                 index_col=None, usecols=None , na_values=['None', 'Null'])
df

Filght_date  Flight_ID Departure_point Arrival_point  \
0       2016-01-03   TU 0712              CMN           TUN   
1       2016-01-13   TU 0757              MXP           TUN   
2       2016-01-16   TU 0214              TUN           IST   
3       2016-01-17   TU 0480              DJE           NTE   
4       2016-01-17   TU 0338              TUN           ALG   
...            ...        ...             ...           ...   
107828  2018-07-05  WKL 0000              TUN           TUN   
107829  2018-01-13   UG 0003              DJE           TUN   
107830  2018-11-07  SGT 0000              TUN           TUN   
107831  2018-01-23   UG 0010              TUN           DJE   
107832  2018-11-13   UG 0002              TUN           DJE   

       Scheduled_departure_time Scheduled_arrival_time STATUS Aircraft_code  \
0           2016-01-03 10:30:00    2016-01-03 12.55.00    ATA     TU 32AIMN   
1           2016-01-13 15:05:00    2016-01-13 16.55.00    ATA     TU 31BIMO   
2           2016-01-16 04:10:00    2016-01-16 06.45.00    ATA     TU 32AIMN   
3           2016-01-17 14:10:00    2016-01-17 17.00.00    ATA     TU 736IOK   
4           2016-01-17 14:30:00    2016-01-17 15.50.00    ATA     TU 320IMU   
...                         ...                    ...    ...           ...   
107828      2018-07-05 23:00:00    2018-07-06 02.00.00    SCH     TU 32AIML   
107829      2018-01-13 08:00:00    2018-01-13 09.00.00    SCH     UG AT7AT7   
107830      2018-11-07 05:00:00    2018-11-07 12.50.00    SCH     TU 736IOK   
107831      2018-01-23 18:00:00    2018-01-23 18.45.00    ATA     TU CR9ISA   
107832      2018-11-13 06:15:00    2018-11-13 07.05.00    SCH     TU CR9ISA   

        Arrival_delay  
0               260.0  
1                20.0  
2                 0.0  
3                 0.0  
4                22.0  
...               ...  
107828            0.0  
107829            0.0  
107830            0.0  
107831            0.0  
107832            0.0  

[107833 rows x 9 columns]

In [3]:
df.head(10)

Filght_date Flight_ID Departure_point Arrival_point  \
0  2016-01-03  TU 0712              CMN           TUN   
1  2016-01-13  TU 0757              MXP           TUN   
2  2016-01-16  TU 0214              TUN           IST   
3  2016-01-17  TU 0480              DJE           NTE   
4  2016-01-17  TU 0338              TUN           ALG   
5  2016-01-17  TU 0283              TLS           TUN   
6  2016-01-18  TU 0514              TUN           BCN   
7  2016-01-18  TU 0716              TUN           ORY   
8  2016-01-18  TU 0752              TUN           FCO   
9  2016-01-18  TU 0996              TUN           NCE   

  Scheduled_departure_time Scheduled_arrival_time STATUS Aircraft_code  \
0      2016-01-03 10:30:00    2016-01-03 12.55.00    ATA     TU 32AIMN   
1      2016-01-13 15:05:00    2016-01-13 16.55.00    ATA     TU 31BIMO   
2      2016-01-16 04:10:00    2016-01-16 06.45.00    ATA     TU 32AIMN   
3      2016-01-17 14:10:00    2016-01-17 17.00.00    ATA     TU 736IOK   
4      2016-01-17 14:30:00    2016-01-17 15.50.00    ATA     TU 320IMU   
5      2016-01-17 16:20:00    2016-01-17 18.15.00    ATA     TU 736IOP   
6      2016-01-18 07:15:00    2016-01-18 09.00.00    ATA     TU 32AIMH   
7      2016-01-18 07:35:00    2016-01-18 09.55.00    ATA     TU 32AIMI   
8      2016-01-18 07:40:00    2016-01-18 09.00.00    ATA     TU 32AIMC   
9      2016-01-18 07:45:00    2016-01-18 09.15.00    ATA     TU 31AIMK   

   Arrival_delay  
0          260.0  
1           20.0  
2            0.0  
3            0.0  
4           22.0  
5           53.0  
6           10.0  
7           15.0  
8           16.0  
9           21.0

In [4]:
df.info() # We only have one numeric variable 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 107833 entries, 0 to 107832
Data columns (total 9 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   Filght_date               107833 non-null  object 
 1   Flight_ID                 107833 non-null  object 
 2   Departure_point           107833 non-null  object 
 3   Arrival_point             107833 non-null  object 
 4   Scheduled_departure_time  107833 non-null  object 
 5   Scheduled_arrival_time    107833 non-null  object 
 6   STATUS                    107833 non-null  object 
 7   Aircraft_code             107833 non-null  object 
 8   Arrival_delay             107833 non-null  float64
dtypes: float64(1), object(8)
memory usage: 7.4+ MB


In [7]:
df.isnull().sum() # There is no missing values

Filght_date                 0
Flight_ID                   0
Departure_point             0
Arrival_point               0
Scheduled_departure_time    0
Scheduled_arrival_time      0
STATUS                      0
Aircraft_code               0
Arrival_delay               0
dtype: int64

In [8]:
df.describe()

Arrival_delay
count  107833.000000
mean       48.733013
std       117.135562
min         0.000000
25%         0.000000
50%        14.000000
75%        43.000000
max      3451.000000

In [16]:
df['Arrival_point'].nunique()

128

In [9]:
from ydata_profiling import ProfileReport

/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [10]:
report = ProfileReport(df)
report.to_file('PR.html')

Export report to file: 100%|██████████| 1/1 [00:00<00:00, 500.99it/s]


In [20]:
# Finding outliers 
Q1 = df['Arrival_delay'].quantile(0.25)
Q3 = df['Arrival_delay'].quantile(0.75)
IQR = Q3 - Q1

# Define the bounds for outliers
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Identifying outliers
outliers = df[(df['Arrival_delay'] < lower_bound) | (df['Arrival_delay'] > upper_bound)]

# Display outliers
print(outliers)


       Filght_date Flight_ID Departure_point Arrival_point  \
0       2016-01-03  TU 0712              CMN           TUN   
17      2016-01-23  TU 0726              MIR           BRU   
38      2016-02-05  TU 0711              TUN           CMN   
46      2016-02-06  TU 0635              ORY           DJE   
56      2016-02-14  TU 0720              TUN           ORY   
...            ...       ...             ...           ...   
103553  2018-12-30  TU 0719              ORY           TUN   
103820  2018-12-22  TU 0282              TUN           TLS   
104782  2018-01-31  UG 1301              MLA           TUN   
104813  2018-12-22  TU 0750              TUN           LYS   
106121  2018-07-29  TU 0791              LHR           TUN   

       Scheduled_departure_time Scheduled_arrival_time STATUS Aircraft_code  \
0           2016-01-03 10:30:00    2016-01-03 12.55.00    ATA     TU 32AIMN   
17          2016-01-23 05:30:00    2016-01-23 08.15.00    ATA     TU 31AIMJ   
38          2016-0

In [21]:
len(outliers)


14608

In [25]:
df.query('Arrival_delay >= 2743') # c'est la 10 plus grande valeur des retards 

Filght_date Flight_ID Departure_point Arrival_point  \
10123  2016-03-20  TU 0209              NTE           TUN   
48878  2017-09-30  TU 0875              GVA           DJE   
50779  2017-11-25  TU 0613              BKO           TUN   
51432  2017-12-11  TU 0250              MIR           NCE   
51435  2017-12-11  TU 0251              NCE           MIR   
59199  2017-08-31  TU 2771              CDG           TUN   
77495  2018-05-13  TU 2588              DJE           LYS   
78095  2018-06-23  TU 0401              ABJ           TUN   
81542  2018-03-22  TU 0611              CZL           TUN   
86009  2018-03-22  TU 0610              TUN           CZL   
99731  2018-12-27  TU 0399              ABJ           TUN   

      Scheduled_departure_time Scheduled_arrival_time STATUS Aircraft_code  \
10123      2016-03-20 10:30:00    2016-03-20 12.50.00    ATA     TU 736IOQ   
48878      2017-09-30 09:25:00    2017-09-30 11.45.00    ATA     TU 32AIMM   
50779      2017-11-25 00:50:00    2017-11-25 05.25.00    ATA     TU 320IMS   
51432      2017-12-11 11:30:00    2017-12-11 13.15.00    ATA     TU 736ION   
51435      2017-12-11 14:05:00    2017-12-11 15.50.00    ATA     TU 736ION   
59199      2017-08-31 10:15:00    2017-08-31 12.45.00    ATA     TU 32AIMM   
77495      2018-05-13 12:00:00    2018-05-13 14.20.00    ATA     TU 32AIML   
78095      2018-06-23 22:40:00    2018-06-24 03.55.00    ATA     TU 320IMW   
81542      2018-03-22 14:25:00    2018-03-22 15.15.00    ATA     TU 320IMT   
86009      2018-03-22 12:45:00    2018-03-22 13.35.00    ATA     TU 320IMT   
99731      2018-12-27 22:40:00    2018-12-28 03.50.00    ATA     TU 320IMU   

       Arrival_delay  
10123         2980.0  
48878         3451.0  
50779         2942.0  
51432         2743.0  
51435         2746.0  
59199         3025.0  
77495         3245.0  
78095         2745.0  
81542         2835.0  
86009         2745.0  
99731         3317.0

Statistiques descriptives des retards de vols : 


Moyenne des retards : 48,73 minutes
    Les vols de Tunisair présentent un retard moyen de près de 49 minutes, indiquant que les retards ne sont pas rares, bien qu'il soit important de noter la dispersion importante autour de cette moyenne.

Médiane des retards : 14 minutes
    La moitié des vols accusent un retard de 14 minutes ou moins, ce qui montre une répartition asymétrique des retards, avec de nombreux vols faiblement en retard.

Premier quartile (Q1) : 0 minute
    Cela signifie qu'au moins 25 % des vols n'ont subi aucun retard.

Troisième quartile (Q3) : 43 minutes
    75 % des vols ont un retard de 43 minutes ou moins, indiquant que la majorité des retards restent dans une plage modérée.

95e percentile : 226 minutes
    95 % des vols ont un retard inférieur à 226 minutes, ce qui montre que les retards extrêmement longs sont rares mais présents.

Écart-type : 117,14 minutes
    L’écart-type relativement élevé reflète la variabilité importante des retards, avec des écarts significatifs par rapport à la moyenne.

Coefficient de variation (CV) : 2,40
    Le CV supérieur à 2 indique une forte dispersion relative des retards, soulignant que la variabilité des retards par rapport à la moyenne est très importante.

Variance : 13,720,74
    La variance confirme l'existence d'une large dispersion des données autour de la moyenne, suggérant une distribution de retards très inégale.


Distribution des retards :

Minimum : 0 minute (certains vols sont à l'heure ou en avance)
Maximum : 3 451 minutes (environ 57 heures) pour le vol le plus en retard. Ce retard extrême suggère la présence d’anomalies dans certains cas.
nterquartile range (IQR) : 43 minutes
Cela indique que la dispersion centrale des retards est concentrée autour de retards modérés, situés entre 0 et 43 minutes pour la majorité des vols.


Top 3 des retards extrêmes : 
  1/  3451 minutes
    Départ : GVA (Genève)
    Arrivée : DJE (Djerba)
  2/ 3317 minutes
    Départ : ABJ (Abidjan)
    Arrivée : TUN (Tunis)
   3/ 3245 minutes
    Départ : DJE (Djerba)
    Arrivée : LYS (Lyon)

Ces valeurs extrêmes représentent des retards dépassant largement les 2 jours, avec des vols partant de différentes destinations telles que Genève (GVA) et Abidjan (ABJ). 
Une attention particulière devrait être portée sur ces itinéraires pour identifier les causes de ces retards .
